tensorflow/python/keras/benchmark/benchmark_util.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Common utils for benchmark."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import timeit
+import numpy as np
+
+from tensorflow.python.keras import callbacks
+from tensorflow.python.keras.benchmark import distribution_util
+
+
+class TimerCallBack(callbacks.Callback):
+  """Callback for logging time in each epoch or batch."""
+
+  def __init__(self):
+    self.times = []
+    self.timer = timeit.default_timer
+    self.startup_time = timeit.default_timer()
+    self.recorded_startup = False
+
+  def on_epoch_begin(self, e, logs):
+    self.epoch_start_time = self.timer()
+
+  def on_epoch_end(self, e, logs):
+    self.times.append(self.timer() - self.epoch_start_time)
+
+  def on_batch_end(self, e, logs):
+    if not self.recorded_startup:
+      self.startup_time = self.timer() - self.startup_time
+      self.recorded_startup = True
+
+
+def measure_performance(model_fn,
+    x=None,
+    y=None,
+    epoch=2,
+    batch_size=32,
+    run_iters=4,
+    optimizer=None,
+    loss=None,
+    metrics=None,
+    verbose=0,
+    num_gpus=0,
+    distribution_strategy='off'):
+  """Run models and measure the performance.
+
+  Arguments:
+    model_fn: Model function to be benchmarked.
+    x: Input data. See `x` in the `fit()` method of `keras.Model`.
+    y: Target data. See `y` in the `fit()` method of `keras.Model`.
+    epoch: Integer. Number of epochs to train the model.
+        If unspecified, `epoch` will default to 2.
+    batch_size: Integer. Number of samples per gradient update.
+        If unspecified, `batch_size` will default to 32.
+    run_iters: Integer. Number of iterations to run the performance measurement.
+        If unspecified, `run_iters` will default to 4.
+    metrics: Lists of metrics to be evaluated by the model during training.
+        See `metrics` in the `compile()` method of  `keras.Model`.
+    verbose: 0, 1, 2. Verbosity mode. See `verbose` in the `fit()`
+        method of `keras.Model`. If unspecified, `verbose` will default to 0.
+    num_gpus: Number of GPUs to run the model.
+    distribution_strategy: Distribution strategies.
+        It could be `tpu`, `multi_worker_mirrored`, `one_device`, `mirrored`,
```

Comment:
Do all the strategies work? If not, how about adding a TODO in line 108 where distribution strategy is created?

Reply:
I couldn't test with the TPU strategy and parameter_server strategy seems to need users to provide TF_CONFIG. So before, I just use it from the model garden team. Maybe I can delete these two and list them as TODO. What do you think?

Reply:
Yeah, makes sense to me.